### PR TITLE
Improve JBMC command line options

### PR DESCRIPTION
**Pull request by peterschrammel**
A first step towards making the JBMC CLI more user-friendly. This includes eg making options default that are always required if running JBMC on a program.

### CODE REVIEW

File: jbmc/src/java_bytecode/java_bytecode_language.h [hunk 1]
```diff
@@ -50,15 +54,17 @@ Author: Daniel Kroening, kroening@kroening.com
...
+  "                                  at level N references are set to null\n" /* NOLINT(*) */ \
   " --java-max-vla-length            limit the length of user-code-created arrays\n"             /* NOLINT(*) */ \
   " --java-cp-include-files          regexp or JSON list of files to load (with '@' prefix)\n"   /* NOLINT(*) */ \
   " --lazy-methods                   only translate methods that appear to be reachable from\n"  /* NOLINT(*) */ \
-  "                                  the --function entry point or main class\n"                 /* NOLINT(*) */ \
+  "                                  the --function entry point or main class (default)\n"       /* NOLINT(*) */ \
   "                                  Note --show-symbol-table/goto-functions/properties output\n"/* NOLINT(*) */ \
   "                                  will be restricted to loaded methods in this case\n"        /* NOLINT(*) */ \
+  " --no-lazy-methods                load and translate all methods given on the command line\n" /* NOLINT(*) */ \
+  "                                  and in --classpath\n"                                       /* NOLINT(*) */ \
```
Comment (mgudemann) [inline jbmc/src/java_bytecode/java_bytecode_language.h]: I'd say the description of `--lazy-methods` should be merged into `--no-lazy-methods`, i.e., "load and translate all ... the default is to translate only the methods that appear to be reachable ..."

File: jbmc/src/jbmc/jbmc_parse_options.cpp [hunk 2]
```diff
@@ -1136,10 +1136,11 @@ void jbmc_parse_optionst::help()
     " --yices                      use Yices\n"
     " --z3                         use Z3\n"
     " --refine                     use refinement procedure (experimental)\n"
-    " --refine-strings             use string refinement (experimental)\n"
+    " --refine-strings             use string refinement (default)\n"
```
Comment (mgudemann) [inline jbmc/src/jbmc/jbmc_parse_options.cpp]: should be removed

**Comment by mgudemann** [comment]
@peterschrammel JBMC unit tests seems to fail, there's probably an implicit assumption on some of the changed defaults for parameters

File: jbmc/src/java_bytecode/java_bytecode_language.h [hunk 3]
```diff
@@ -31,12 +31,16 @@ Author: Daniel Kroening, kroening@kroening.com
   "(disable-uncaught-exception-check)" \
   "(throw-assertion-error)" \
   "(java-assume-inputs-non-null)" \
-  "(java-throw-runtime-exceptions)" \
-  "(java-max-input-array-length):" \
-  "(java-max-input-tree-depth):" \
+  "(java-throw-runtime-exceptions)" /* will go a away */ \
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_bytecode_language.h]: a away -> away

**Comment by peterschrammel** [comment]
> JBMC unit tests seems to fail

Fixing that right now

File: jbmc/regression/jbmc/coreModels/test.desc [hunk 4]
```diff
@@ -3,7 +3,7 @@ test.class
 --show-symbol-table --cp ../../../src/java_bytecode/library/core-models.jar:.
 ^EXIT=0$
 ^SIGNAL=0$
-^Symbol\s*\.*\: java\:\:org\.cprover\.CProver\.\<clinit\>\:\(\)V$
+^Symbol\s*\.*\: java\:\:java.lang.Throwable\.\<clinit\>\:\(\)V$
 --
 --
-tests that the core models are  being loaded by checking if the static initializer for the CProver class was
+tests that the core models are  being loaded by checking if the static initializer for the Throwable class was loaded
```
Comment (kroening) [inline jbmc/regression/jbmc/coreModels/test.desc]: Remove double space on the occasion?

**Comment by kroening** [comment]
The key change that needs to happen to the JBMC command line is to use the java convention of giving a class name, instead of a file name of a class file.

**Comment by peterschrammel** [comment]
> The key change that needs to happen to the JBMC command line is to use the java convention of giving a class name, instead of a file name of a class file.

Yes, but that's a second step after this one (requires modifying *all* regression tests).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: fed3e13564f093da0b0f38c7919bf6f930a0208e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78364616

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: d139adcc5c976cc8e50b6253cdf36ddcf0860d61).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78380492

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 0a10bf372d961bc3db72942adf2657f611616ce6).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78388715

**Comment by peterschrammel** [comment]
@smowton, @mgudemann, TG bump finally passes.

**Review by mgudemann — APPROVED** [review]
LGTM

**Review by smowton — APPROVED** [review]
Looks reasonable to me apart from the concern about tests above, so approving on the assumption that is dealt with one way or another.

**Comment by peterschrammel** [comment]
@smowton, addressed in the TG bump.

**Pull request merged**